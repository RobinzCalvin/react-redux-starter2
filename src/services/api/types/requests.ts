import { IFlatFormProperties } from 'shared/types/models';

export interface ITravelOrderRequest {
  attributes: IFlatFormProperties;
  notify: boolean;
  description: string;
  location: number;  // area id only - don't need a city id
  category: number;
  coord_from_lng: number;
  coord_from_lat: number;
  coord_to_lng: number;
  coord_to_lat: number;
}
