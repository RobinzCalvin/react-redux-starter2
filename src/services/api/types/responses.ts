export interface IGithubAPIResponse<T> {
  total_count: number;
  items: T[];
}

export type SearchUserResponse = IGithubAPIResponse<IServerUser>;

export type SearchRepositoriesResponse = IGithubAPIResponse<IServerRepository>;

export interface IServerUser {
  id: number;
  login: string;
  avatar_url: string;
  html_url: string;
}

export interface IDetailedServerUser extends IServerUser {
  followers: number;
  following: number;
  public_repos: number;
  name: string;
  location: null | string;
}

export interface IServerRepository {
  id: number;
  open_issues: number;
  forks: number;
  stargazers_count: number;
  description: string;
  updated_at: string;
  name: string;
  html_url: string;
  language: string;
  owner: IServerUser;
}
