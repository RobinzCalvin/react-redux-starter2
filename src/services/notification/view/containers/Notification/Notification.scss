.notification {
  $r: &;

  position: fixed;
  top: 0;
  left: 50%;
  z-index: 1002;
  display: flex;
  justify-content: flex-start;
  padding: 0.79rem 0.71rem;
  max-width: 42.857rem;
  border-radius: 0 0 2px 2px;
  color: white;
  font-weight: 500;
  font-size: 0.86rem;
  transform: translateX(-50%);
  animation: fadeInOut 10s ease;

  &__holder {
    display: flex;

    &__content {
      z-index: 1001;
    }
  }

  &_kind {
    &_error {
      background-color: #ef4c4c;
    }

    &_info {
      background-color: #44ba39;
    }
  }

  &__icon {
    width: 0.93rem;
    height: 0.93rem;
  }
}

@keyframes fadeInOut {
  0% {
    opacity: 0;
  }

  25% {
    opacity: 1;
  }

  50% {
    opacity: 1;
  }

  75% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}
