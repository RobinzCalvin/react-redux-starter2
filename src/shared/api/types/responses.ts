export interface IAreaResponse {
  city: number;
  display_name: string;
  name: string;
  point: string;
}

export interface ICityResponse {
  areas: IAreaResponse[];
  name: string;
  id: number;
}
