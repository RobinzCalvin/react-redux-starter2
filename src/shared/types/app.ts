import { ReactElement } from 'react';
import { RouteProps } from 'react-router';
import { Store, Reducer, ActionCreator, Action } from 'redux';

import { SagaIterator } from 'redux-saga';

import { namespace as CategorySelectNamespace } from 'features/categorySelect/entry';
import { namespace as LocationSelectNamespace } from 'features/locationSelect/entry';
import { namespace as SearchRepositoriesNamespace } from 'features/searchRepositories/entry';
import { namespace as DynamicFieldsNamespace } from 'features/dynamicFields/entry';
import { Namespace as HomeModuleNamespace } from '../../modules/OrderForm/OrderForm';

import Api from '../api/Api';

export abstract class Module<C = any> {
  public components?: C; // available componens to pass in other modules

  protected _store: Store<IAppReduxState> | null = null;
  protected _deps: IDependencies | null = null;

  protected extraComponents?: { [key: string]: React.ReactElement<any> | React.ComponentClass<any> | null; };

  public set store(store: Store<IAppReduxState>) {
    this._store = store;
  }

  public set dependencies(value: IDependencies) {
    this._deps = value;
  }

  protected get deps(): IDependencies | null {
    return this._deps || null;
  }

  public getRoutes?(): ReactElement<RouteProps> | Array<ReactElement<RouteProps>>;
  public getReduxEntry?(): IReduxEntry;

  public setExtraComponent(key: keyof C, component: React.ReactElement<any>): void {
    if (this.extraComponents) {
      this.extraComponents[key] = component;
    } else {
      throw new Error('Cannot set module extra component: no requirements found for extra component');
    }
  }
}

export interface IAction {
  type: string;
  payload?: any;
}

export interface IAppData {
  modules: Module[];
  store: Store<IAppReduxState>;
}

export interface IDependencies {
  api: Api;
}

export type IDictionary<T, S extends string = string> = {
  [key in S]: T;
};

export interface IReduxEntry {
  reducers?: {[key in keyof IAppReduxState]?: Reducer<IAppReduxState[key]>};
  sagas?: RootSaga[];
}

// TODO: add averload for aptional A, S, C
export interface IFeatureEntry<
  C extends IDictionary<ReactComponent<any>, keyof C> | void,
  A extends IDictionary<ActionCreator<Action>, keyof A> | void,
  S extends IDictionary<(state: any, ...args: any[]) => any, keyof S> | void,
> extends IReduxEntry {
  actions: A;
  selectors: S;
  containers: C;
}

export interface IAppReduxState {
  categorySelect: CategorySelectNamespace.IReduxState;
  locationSelect: LocationSelectNamespace.IReduxState;
  dynamicFields: DynamicFieldsNamespace.IReduxState;
  orderForm: HomeModuleNamespace.IReduxState;
  searchRepositories: SearchRepositoriesNamespace.IReduxState;
}

export type Diff<T extends string, U extends string> =
  ({[P in T]: P } & {[P in U]: never } & { [x: string]: never })[T];
export type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;

export type RootSaga = (deps: IDependencies) => () => SagaIterator;

export type BundleLoader<T> = (callback: (bundle: T) => void) => void;

export type Lang = 'en' | 'he';

export type ReactComponent<P> = React.ComponentClass<P> | React.StatelessComponent<P>;
