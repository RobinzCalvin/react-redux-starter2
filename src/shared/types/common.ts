export interface IPaginationState {
  page: number;
  totalPages: number;
}

export type NotificationKind = 'error' | 'info';

export interface INotification {
  kind: NotificationKind;
  text: string;
}

type ValueFormatter<T> = (x: T) => any;
export type KeysToValuesFormattersMap<T> = Partial<{[K in keyof T]: ValueFormatter<T[K]>}>;
