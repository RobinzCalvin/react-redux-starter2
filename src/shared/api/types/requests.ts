import { IAttributes } from 'shared/types/models';

export interface IOrderFormRequest {
  attributes: IAttributes;
  notify: boolean;
  description: string;
  location: number;  // area id only - don't need a city id
  category: number;
  coord_from_lng: number;
  coord_from_lat: number;
  coord_to_lng: number;
  coord_to_lat: number;
}
