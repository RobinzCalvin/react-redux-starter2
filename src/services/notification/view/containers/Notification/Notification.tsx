import * as React from 'react';
import { connect } from 'react-redux';
import block from 'bem-cn';

import { IAppReduxState } from 'shared/types/app';
import { INotification } from 'shared/types/ui';

import { selectors } from '../../../redux';
import './Notification.scss';

interface IStateProps {
  notification: INotification | null;
}

function mapState(state: IAppReduxState): IStateProps {
  return {
    notification: selectors.selectNotification(state),
  };
}

type IProps = IStateProps;

const b = block('notification');

class Notification extends React.PureComponent<IProps> {

  public render() {
    const { notification } = this.props;
    return (
      notification && this.renderContent(notification)
    );
  }

  private renderContent(notification: INotification) {
    const { kind, text } = notification;
    return (
      <div className={b({ kind })}>
        <div className={b('text')}>
          {text}
        </div>
      </div>
    );
  }
}

export default connect(mapState)(Notification);
