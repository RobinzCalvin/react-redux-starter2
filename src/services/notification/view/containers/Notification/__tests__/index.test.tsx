import React from 'react';
import { mount, shallow } from 'enzyme';
import { INotification } from 'shared/types/common';
import { Notification } from '../Notification';

describe('Notification component', () => {
  const notification: INotification = {
    kind: 'error',
    text: 'notification text',
  };
  const component = mount(<Notification notification={notification} />);

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render notification text', () => {
    expect(component.find('.notification__text').text()).toBe(notification.text);
  });

  it('should not render if no notification provided', () => {
    const emptyNotification = shallow(<Notification notification={null} />);
    expect(emptyNotification.html()).toBeNull();
  });
});
