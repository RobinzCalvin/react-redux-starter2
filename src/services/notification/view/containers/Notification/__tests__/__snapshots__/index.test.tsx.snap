// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component should match snapshot 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    notification text
  </div>
</div>
`;
